core/primitives/src/stateless_validation.rs

```diff
@@ -24,16 +25,31 @@ type SignatureDifferentiator = String;
 #[derive(Debug, Clone, PartialEq, Eq, BorshSerialize, BorshDeserialize)]
 pub struct EncodedChunkStateWitness(Box<[u8]>);
 
+pub type ChunkStateWitnessSize = usize;
+
 impl EncodedChunkStateWitness {
-    pub fn encode(witness: &ChunkStateWitness) -> Self {
-        Self(borsh::to_vec(witness).expect("borsh serialization should not fail").into())
+    /// Borsh-serialize and compress state witness.
+    /// Returns encoded witness along with the raw (uncompressed) witness size.
+    pub fn encode(witness: &ChunkStateWitness) -> std::io::Result<(Self, ChunkStateWitnessSize)> {
+        const STATE_WITNESS_COMPRESSION_LEVEL: i32 = 3;
+        let borsh_bytes = borsh::to_vec(witness)?;
+        Ok((
+            Self(zstd::encode_all(borsh_bytes.as_slice(), STATE_WITNESS_COMPRESSION_LEVEL)?.into()),
+            borsh_bytes.len(),
+        ))
     }
 
-    pub fn decode(&self) -> Result<ChunkStateWitness, std::io::Error> {
-        ChunkStateWitness::try_from_slice(&self.0)
+    /// Decompress and borsh-deserialize encoded witness bytes.
+    /// Returns decoded witness along with the raw (uncompressed) witness size.
+    pub fn decode(&self) -> std::io::Result<(ChunkStateWitness, ChunkStateWitnessSize)> {
+        // We want to limit the size of decompressed data to address "Zip bomb" attack.
+        const MAX_WITNESS_SIZE: usize = 64_000_000;
```

Comment:
It's a bit risky to limit state witness size to 64MB, what if there's some corner case that would blow up the size above that? I have flashbacks from https://github.com/near/nearcore/pull/10615 x.x

AFAIK we have a 500MB limit for incoming network messages, maybe that'd be a bit safer? But at the end of the day we have to choose some value, so maybe 64MB is okay.

Reply:
Good point, I will change that to match the network message size limit.

Reply:
If the network messages are limited to 500M, aren't these witnesses already limited to that amount?

Reply:
Compressed state witness is indeed limited by the network message size, but at this point we haven't verified that data yet, so it can be malicious which could result in large memory usage when decompressed and crash the node. That is why we also need to explicitly limit the size of decompressed data here.

Reply:
@jancionear the limit was updated in 94a406b02a77fc780b8824bdd334e095dd2dd6cb